scripts.Rmd

```diff
@@ -0,0 +1,255 @@
+# `drake` and script-based workflows {#scripts}
+
+```{r, message = FALSE, warning = FALSE, echo = FALSE}
+knitr::opts_knit$set(root.dir = fs::dir_create(tempfile()))
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
+```
+
+```{r, message = FALSE, warning = FALSE, echo = FALSE}
+library(drake)
+library(glue)
+library(purrr)
+library(rlang)
+library(tidyverse)
+invisible(drake_example("script-based-workflows", overwrite = TRUE))
+tmp <- file.copy(
+  list.files("script-based-workflows/R/",
+             pattern = "*.R", full.names = TRUE),
+  ".",
+  recursive = TRUE)
+tmp <- file.copy("script-based-workflows/raw_data.xlsx", ".")
+tmp <- file.copy("script-based-workflows/report.Rmd", ".")
+dir.create("data")
+```
+
+`drake` has been built with the firm belief that the best workflows rely on approaching problems from a function perspective. This idea is supported in "Good enough practices in scientific computing", that approaching your project by writing modular code increases the end product.
```

Comment:
I just realized we might want to

1. Streamline the first sentence, and
2. Talk a little about functions in plans first.

Maybe something like this:

## Function-oriented workflows

`drake` wants you to write functions for data analysis. Functions break down complicated ideas into manageable pieces.

```r
# R/functions.R
get_data <- function(file) {
  readxl::read_excel(file)
}

munge_data <- function(raw_data) {
  raw_data %>%
    mutate(Species = forcats::fct_inorder(Species))
}

fit_model <- function(munged_data) {
  lm(Sepal.Width ~ Petal.Width + Species, munged_data)
}
```

`get_data()`, `munge_data()`, and `fit_model()` are a special shorthand that makes the plan easier to read.

```r
# R/plan.R
drake_plan(
  raw_data = get_data(file_in("raw_data.xlsx")),
  munged_data = munge_data(raw_data),
  model = fit_model(munged_data)
)
```

To invoke `drake`, we load the supporting scripts and call `make()`.

```r
source("R/packages.R")
source("R/functions.R")
source("R/plan.R")
```

This function-oriented approach is elegant, powerful, testable, scalable, and maintainable. However, it can be challenging to convert pre-existing traditional script-based analyses to function-oriented `drake`-powered workflows. The stopgap in this chapter quickly retrofits `drake` and functions onto existing projects without needing to change the original underlying scripts.